sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -300,97 +300,71 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
  */
 object ColumnPruning extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case a @ Aggregate(_, _, e @ Expand(projects, output, child))
-      if (e.outputSet -- a.references).nonEmpty =>
-      val newOutput = output.filter(a.references.contains(_))
-      val newProjects = projects.map { proj =>
-        proj.zip(output).filter { case (e, a) =>
+    // Prunes the unused columns from project list of Project/Aggregate/Window/Expand
+    case p @ Project(_, p2: Project) if (p2.outputSet -- p.references).nonEmpty =>
```

Comment:
I think `CollapseProject` can also cover this case, it merges the inner project list into the outer one, so unused columns in inner project list are also removed.

But this rule is useful when `CollapseProject` can't be applied, i.e. there are non-deterministic overlapped expressions. BTW I think the order doesn't matter here, this rule can run before or after `CollapseProject`, as it handles cases `CollapseProject` can't handle.

So this LGTM